src/Grpc.AspNetCore.Server/Internal/GrpcProtocolHelpers.cs

```diff
@@ -207,7 +207,21 @@ public static void SetStatusTrailers(HttpResponse response, Status status)
         {
             // Use SetTrailer here because we want to overwrite any that was set earlier
             SetTrailer(response, GrpcProtocolConstants.StatusTrailer, status.StatusCode.ToTrailerString());
-            SetTrailer(response, GrpcProtocolConstants.MessageTrailer, !string.IsNullOrEmpty(status.Detail) ? status.Detail : null);
+
+            string escapedDetail;
+            if (!string.IsNullOrEmpty(status.Detail))
+            {
+                // https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#responses
+                // The value portion of Status-Message is conceptually a Unicode string description of the error,
+                // physically encoded as UTF-8 followed by percent-encoding.
+                escapedDetail = PercentEncodingHelpers.PercentEncode(status.Detail);
+            }
+			else
```

Comment:
why not just declare it as null and set it to a value in the if block

Reply:
Because this will force it to null if it was empty, and avoid writing a header with an empty value.

Reply:
I guess what you suggest does the same thing. I suppose I like code that forces each path of the `if` to make a choice 🤷‍♂

Reply:
if you want to keep empty, just initialize escapeDetail = status.Detail. Also, nit: tabs.